### PR TITLE
Re[CL:Incentives] Add further tests to uptime accumulator update logic

### PR DESCRIPTION
**Pull request by stackman27**
Closes: https://github.com/osmosis-labs/osmosis/issues/4339

## What is the purpose of the change

Opening this PR to merge directly to main and resolve merge conflicts in https://github.com/osmosis-labs/osmosis/pull/4353

## Brief Changelog

n/a

## Testing and Verifying

*(Please pick one of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added unit test that validates ...*
  - *Added integration tests for end-to-end deployment with ...*
  - *Extended integration test for ...*
  - *Manually verified the change by ...*

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

**Review by AlpinYukseloglu — APPROVED** [review]
We should change all uses of `UpTime` to `Uptime` but other than that LGTM – good work!

**Comment by stackman27** [comment]
> We should change all uses of `UpTime` to `Uptime` but other than that LGTM – good work!

hi @AlpinYukseloglu little confused uses of uptime to uptime?

**Comment by czarcas7ic** [comment]
> > We should change all uses of `UpTime` to `Uptime` but other than that LGTM – good work!
> 
> hi @AlpinYukseloglu little confused uses of uptime to uptime?

He is referring to the capitalization

File: x/concentrated-liquidity/incentives_test.go [hunk 1]
```diff
@@ -430,6 +462,106 @@ func (s *KeeperTestSuite) TestCalcAccruedIncentivesForAccum() {
...
+			poolId:              defaultPoolId,
+			accumUptime:         types.SupportedUptimes[0],
+			qualifyingLiquidity: sdk.NewDec(100),
+			timeElapsed:         time.Hour,
+
+			poolIncentiveRecords: []types.IncentiveRecord{incentiveRecordOne, incentiveRecordOneWithDifferentMinUpTime},
+
+			expectedResult: sdk.DecCoins{
+				// We expect both incentive record to qualify
+				expectedIncentives(incentiveRecordOne.IncentiveDenom, incentiveRecordOne.EmissionRate, time.Hour, sdk.NewDec(100)),
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:514]: The comment here says that we expect both to qualify but we only run one of the two for expected incentives.

File: x/concentrated-liquidity/incentives_test.go [hunk 2]
```diff
@@ -430,6 +462,106 @@ func (s *KeeperTestSuite) TestCalcAccruedIncentivesForAccum() {
...
+
+			expectedResult: sdk.DecCoins{
+				// We expect both incentive record to qualify
+				expectedIncentives(incentiveRecordOneWithDifferentStartTime.IncentiveDenom, incentiveRecordOne.EmissionRate, time.Hour, sdk.NewDec(100)),
+				expectedIncentives(incentiveRecordOneWithDifferentDenom.IncentiveDenom, incentiveRecordOne.EmissionRate, time.Hour, sdk.NewDec(100)),
+			},
+			expectedIncentiveRecords: []types.IncentiveRecord{
+				// We only charge the first incentive record because the second minUpTime hasn't been hit yet
+				chargeIncentive(incentiveRecordOneWithDifferentStartTime, time.Hour),
+				chargeIncentive(incentiveRecordOneWithDifferentDenom, time.Hour),
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:500]: This says we only charge the first incentive record but we charge both

File: x/concentrated-liquidity/incentives_test.go [hunk 3]
```diff
@@ -430,6 +462,106 @@ func (s *KeeperTestSuite) TestCalcAccruedIncentivesForAccum() {
...
+
+			poolIncentiveRecords: []types.IncentiveRecord{incentiveRecordOne, incentiveRecordOneWithDifferentDenom},
+
+			expectedResult: sdk.DecCoins{
+				// We expect both incentive record to qualify
+				expectedIncentives(incentiveRecordOne.IncentiveDenom, incentiveRecordOne.EmissionRate, time.Hour, sdk.NewDec(100)),
+				expectedIncentives(incentiveRecordOneWithDifferentDenom.IncentiveDenom, incentiveRecordOne.EmissionRate, time.Hour, sdk.NewDec(100)),
+			},
+			expectedIncentiveRecords: []types.IncentiveRecord{
+				// We only charge the first incentive record because the second minUpTime hasn't been hit yet
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:537]: This comment is wrong

File: x/concentrated-liquidity/incentives_test.go [hunk 4]
```diff
@@ -430,6 +462,106 @@ func (s *KeeperTestSuite) TestCalcAccruedIncentivesForAccum() {
...
+		"four incentive records with only two eligilbe for emitting incentives": {
+			poolId:              defaultPoolId,
+			accumUptime:         types.SupportedUptimes[0],
+			qualifyingLiquidity: sdk.NewDec(100),
+			timeElapsed:         time.Hour,
+
+			poolIncentiveRecords: []types.IncentiveRecord{incentiveRecordOne, incentiveRecordOneWithDifferentStartTime, incentiveRecordOneWithDifferentDenom, incentiveRecordOneWithDifferentMinUpTime},
+
+			expectedResult: sdk.NewDecCoins(
+				// We expect both incentive record to qualify
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:552]: What do we mean "both" when there is four? Also, if there are three expected records, shouldn't there be at least three expected incentives?

File: x/concentrated-liquidity/incentives_test.go [hunk 5]
```diff
@@ -430,6 +462,106 @@ func (s *KeeperTestSuite) TestCalcAccruedIncentivesForAccum() {
...
+				expectedIncentives(incentiveRecordOne.IncentiveDenom, incentiveRecordOne.EmissionRate.Add(incentiveRecordOneWithDifferentStartTime.EmissionRate), time.Hour, sdk.NewDec(100)),
+				expectedIncentives(incentiveRecordOneWithDifferentDenom.IncentiveDenom, incentiveRecordOne.EmissionRate, time.Hour, sdk.NewDec(100)),
+			),
+			expectedIncentiveRecords: []types.IncentiveRecord{
+				// We only charge the first three incentive record because the second minUpTime hasn't been hit yet
+				chargeIncentive(incentiveRecordOne, time.Hour),
+				chargeIncentive(incentiveRecordOneWithDifferentStartTime, time.Hour),
+				chargeIncentive(incentiveRecordOneWithDifferentDenom, time.Hour),
+				incentiveRecordOneWithDifferentMinUpTime, // this uptime hasn't hit yet so we donot have to charge incentive
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:561]:
```suggestion
				incentiveRecordOneWithDifferentMinUpTime, // this uptime hasn't hit yet so we do not have to charge incentive
```

Comment (czarcas7ic) [reply x/concentrated-liquidity/incentives_test.go:561]: Why do we call charge incentive here when the comment says we don't charge it?

**Comment by stackman27** [comment]
fixed all the comments @czarcas7ic

File: x/concentrated-liquidity/incentives_test.go (hunk 4 above)
Comment (stackman27) [reply x/concentrated-liquidity/incentives_test.go]: fixed

**Pull request merged**